server/audit_math/shim.py

```diff
@@ -0,0 +1,130 @@
+"""
+
+import logging
+import math
+from typing import Any
+from decimal import Decimal
+from athena.audit import Audit  # type: ignore
+
+
+def make_election(risk_limit, p_w: float, p_r: float) -> Any:
+    """
+    Transform fractional shares to an athena Election object.
+
+    Inputs:
+        risk_limit      - the risk-limit for this audit
+        p_w             - the fraction of vote share for the winner
+        p_r             - the fraction of vote share for the loser / runner-up
+    """
+
+    # calculate the undiluted "two-way" share of votes for the winner
+    p_wr = p_w + p_r
+    p_w2 = p_w / p_wr
+
+    contest_ballots = 100000
+    winner = int(contest_ballots * p_w2)
+    loser = contest_ballots - winner
+
+    contest = {
+        "contest_ballots": contest_ballots,
+        "tally": {"A": winner, "LOSER": loser},
+        "num_winners": 1,
+        "reported_winners": ["A"],
+        "contest_type": "PLURALITY",
+    }
+
+    contest_name = "ArloContest"
+    election = {
+        "name": "ArloElection",
+        "total_ballots": contest_ballots,
+        "contests": {contest_name: contest},
+    }
+
+    audit = Audit("minerva", risk_limit)
+    audit.add_election(election)
+    audit.load_contest(contest_name)
+
+    return audit
+
+
+def minerva_sample_sizes(
+    alpha: Decimal,
+    p_w: Decimal,
+    p_r: Decimal,
+    sample_w: int,
+    sample_r: int,
+    p_completion: float,
+) -> int:
+    """
+    Return Minerva round size based on completion probability, assuming the election outcome is correct.
+    TODO: refactor to pass in integer vote shares to allow more exact calculations, incorporate or
+    track round schedule over time, and handle sampling without replacement.
+
+    Inputs:
+        risk_limit      - the risk-limit for this audit
+        p_w             - the fraction of vote share for the winner
+        p_r             - the fraction of vote share for the loser
+        sample_w        - the number of votes for the winner that have already
+                          been sampled
+        sample_r        - the number of votes for the runner-up that have
+                          already been sampled
+        p_completion    - the desired chance of completion in one round,
+                          if the outcome is correct
+
+    Outputs:
+        sample_size     - the expected sample size for the given chance
+                          of completion in one round
+
+    >>> minerva_sample_sizes(0.1, 0.6, 0.4, 56, 56, 0.7)
+    326
+
+    # FIXME: check this
+    >>> minerva_sample_sizes(0.1, 0.6, 0.4, 0, 0, 0.7)
+    111
+    >>> minerva_sample_sizes(0.1, 0.6, 0.4, 0, 0, 0.9)
+    179
+    """
+
+    risk_limit = float(alpha)
+    p_w = float(p_w)  # type: ignore
+    p_r = float(p_r)  # type: ignore
+
+    # calculate the undiluted "two-way" share of votes for the winner
+    p_wr = p_w + p_r
+    p_w2 = p_w / p_wr
+
+    audit = make_election(risk_limit, p_w, p_r)  # type: ignore
```

Comment:
So it looks like this is the major performance hit we're taking. Because we're calling this function 4 times per sample_size generation, we're rebuilding the distribution which is slow. So we ought to pass in a list of quantiles rather than calling this function once per quantile.

Reply:
I've pushed a much more straightforward version of the interface code, which uses a single AthenaAudit instance, rather than borrowing from the old BRAVO approach, and eliminates shim.py
But it turns out that there is very little speedup for the first round, which is the only time we're currently calling the code.

The big speedup comes from the concurrent change in athena 0.7.9 to use a very good approximation to the round size for margins under 1.5%.  Above that, on my machine, the total time is about 5 seconds, so Arlo / Heroku should not really notice much slowdown any more.

The slow parts now are risk calculations. which for hundreds of thousands of ballots in a 2nd round can take over a minute on the first round for a 350000 ballot sample, and 90 seconds for 500,000, but is more like 10 seconds for 100,000.  They are slower in the second round.

It looks like a factor of 10x improvement is on the way for calculate_risk though!

Reply:
I've pushed a much more straightforward version of the interface code, which uses a single AthenaAudit instance, rather than borrowing from the old BRAVO approach, and eliminates shim.py
But it turns out that there is very little speedup for the first round, which is the only time we're currently calling the code.

The big speedup comes from the concurrent change in athena 0.7.9 to use a very good approximation to the round size for margins under 1.5%.  Above that, on my machine, the total time is about 5 seconds, so Arlo / Heroku should not really notice much slowdown any more.

The slow parts now are risk calculations. which for hundreds of thousands of ballots in a 2nd round can take over a minute on the first round for a 350000 ballot sample, and 90 seconds for 500,000, but is more like 10 seconds for 100,000.  They are slower in the second round.

It looks like a factor of 10x improvement is on the way for calculate_risk though!